### PR TITLE
Fix list activity notifications

### LINKED ISSUE #11782: Difference in class activity notifications after full facility import
## Observed behavior
In a [recent test](https://github.com/learningequality/kolibri/pull/11765#issuecomment-1906241734) I noticed some differences between class activity notifications displayed between two devices after the full facility import. Today I could not replicate the 'need help' status disappearance after full facility import from Win11 (0.16b10) to Win7 (0.16b13), howver, the class activity notifications continue not to display on the source (Win11) device anymore. The last syncing activity on that Win11 server was 5-6 days ago.

## Errors and logs
[Logs and dbs](https://drive.google.com/drive/folders/1vQlYj-o4Hkcy-GmZnCNQDF8pqADk3nSV?usp=drive_link) for both servers.

## Expected behavior
Consistent display of class activity on both devices

## User-facing consequences
Confusing and inconsistent information presented to the coach.

## Steps to reproduce
Import the full facility (with some class interactions and notifications) from one server to another. 

## Context

 * Kolibri version:  Windows 11 (0.16b10) and Windows 7 (0.16b13)

### PR DESCRIPTION
**Pull request by AlexVelezLl**
## Summary

There have been a large number of inconsistencies between generating notifications on an original device, and creating notifications when a device is importing a facility from another device. Mainly because the notifications are not created in chronological order, but are created as the log records are imported. Which leads to the main assumption that notifications are created in a chronological order being incorrect. This PR fixes the following:
* In the ClassroomNotifications API, Filter and return notifications in chronological order instead of sorting them by id.
  * Updated the before and after params to receive timestamps instead of an id.
  * A new desc index was created on the `timestamp` field of the `learnerprogressnotification` table.
  * A new `ClassroomNotificationsFilter` class was created to update the way to filter records to use filtersets.
* Until now, if a notification already existed, creating/updating the notification generated by other resources/attempts was avoided. But now the notifications are updated in the following cases:
  * The lesson completed notification is updated if a resource is processed that has a greater completion_time than the existing notification had. So we always keep the latest resource completion_time as the lesson completion_time.
  * (for exercise resources) The resource started notification is updated if an attempt is processed that has a lower start_time than the existing notification had. So we always keep the earliest attempt start_time as the resource start_time.
  * The lesson started notification is updated if a resource is processed that has a lower start_time than the existing notification had. So we always keep the earliest resource start_time as the lesson start_time.
  * The help needed notification is updated if the student has newer attempts and keeps having incorrect interactions. So we always keep the timestamp of the most recent attempt that has an incorrect interaction. (This is also a change for the original device).
* Avoid creating notifications when a content summary log is created (in batch) for an exercise resource, as the resource started time for exercise resources is determined by the first attempt and not by the content summary log.
* Now the resource completion time for exercise resources is determined by the earliest attempt completion_time logged in the MasteryLog, rather than using the content summary log end_time.
* Until now, completion notifications were avoided if the contentsummarylog of a resource had progress <1. But this was not true in cases where the student clicked on "try again" and the progress was = 0, but the notification should still be created when importing. So now the summarylog progress and completion_time are taken into account to determine whether or not to:
  * Create a resource completed notification.
  * Determine if all lesson resources are completed, to create a lesson completed notification (this last one is a bug that could also happen on the original device).
* Some refactor and code reuse to avoid many parts of repeated code in the notifications api.
* In the frontend, in the coachnotifications module:
  * the list of notifications was not properly being sorted.
  * We now sort `notifications`, `allNotifications` by timestamp.
  * If the first load of notifications returned just `Answered` notifications, an empty notifications list was rendered without option for the user to see more notifications. So now we load more notifications until the user could see the latests notifications.

## References

Closes #11782.

## Reviewer guidance

Set up a device with multiiple lessons/resources/quizes and play with it to have different types of notifications. Then import the facility to a new device and check that notificatoins match.

----

## Testing checklist

- [x] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests

## PR process

- [ ] PR has the correct target branch and milestone
- [ ] PR has 'needs review' or 'work-in-progress' label
- [ ] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

### CODE REVIEW

File: kolibri/core/logger/api.py [hunk 1]
```diff
@@ -823,8 +823,7 @@ def _update_summary_log(
         summarylog = ContentSummaryLog.objects.get(
             content_id=sessionlog.content_id, user=user
         )
-        was_complete = summarylog.progress >= 1
```
Comment (rtibbles) [inline kolibri/core/logger/api.py]: I think I would prefer we not make this change - as I think it makes sense for the summary log's completion timestamp to be updated to the latest completion.

In the only circumstance where this can happen, a practice quiz with multiple attempts, we still have the complete history of completion_timestamps on the MasteryLogs associated, but the final ContentSummaryLog completion_timestamp is the most recent (this is different to the notification though, as there we are noting the _first_ time it was completed).

File: kolibri/plugins/coach/api.py [hunk 2]
```diff
@@ -192,14 +192,18 @@ def get_queryset(self):
         notifications_query = self.apply_group_filter(notifications_query)
         after = self.check_after()
```
Comment (rtibbles) [inline kolibri/plugins/coach/api.py]: It feels like it makes sense to me for the `after` value to no longer be an id, but rather be a timestamp, that way we can avoid the indirection you've had to deal with below.

File: kolibri/plugins/coach/api.py [hunk 3]
```diff
@@ -212,12 +216,15 @@ def get_queryset(self):
         limit = self.check_limit()
         if before and limit:
```
Comment (rtibbles) [inline kolibri/plugins/coach/api.py]: Likewise for before - I feel like it makes sense to change the whole semantics of the API endpoint to be focused on timestamps rather than ids.

File: kolibri/plugins/coach/assets/src/modules/coachNotifications/index.js [hunk 4]
```diff
@@ -87,7 +91,7 @@ export default {
     },
     moreNotificationsForClass(store, params) {
       const classroomId = store.state.currentClassroomId;
-      const lastNotification = store.state.notifications.slice(-1)[0];
```
Comment (rtibbles) [inline kolibri/plugins/coach/assets/src/modules/coachNotifications/index.js]: If our data is ordered by timestamp, do we need to make this change?

File: kolibri/plugins/coach/assets/src/views/common/notifications/ActivityList.vue [hunk 5]
```diff
@@ -154,9 +154,6 @@
           const params = {
             ...this.filterParam,
           };
-          if (this.notifications.length) {
-            params.before = this.notifications.slice(-1)[0].id;
```
Comment (rtibbles) [inline kolibri/plugins/coach/assets/src/views/common/notifications/ActivityList.vue:158]: Why has this been removed?

**Review by rtibbles — CHANGES_REQUESTED** [review]
A good start, but I think we can make the changeover to timestamp a bit more systematically.

One other thing that would be important to investigate in this changeover, is whether we need to add a db_index to the timestamp field: https://github.com/learningequality/kolibri/blob/develop/kolibri/core/notifications/models.py#L117 the primary key (id) is indexed by default, so querying solely by timestamp may be considerably less performant.

File: kolibri/core/logger/api.py (hunk 1 above)
Comment (AlexVelezLl) [reply kolibri/core/logger/api.py]: All right, I will check the MasteryLogs and see how we can query it to get the first timestamp

File: kolibri/plugins/coach/api.py (hunk 2 above)
Comment (AlexVelezLl) [reply kolibri/plugins/coach/api.py]: Oh I did it to keep the same API :sweat_smile:. And also didnt found any example of how do we send timestamps as filters to the rest api. Do we send a unix timestamp from the frontend? Or the same ISO date format we send from the backend? And after that use something like `after = datetime.datetime.fromisoformat(notifications_after)`? If there is any example of other place where we do something similar woul appreciate it :open_hands:.

File: kolibri/plugins/coach/assets/src/modules/coachNotifications/index.js (hunk 4 above)
Comment (AlexVelezLl) [reply kolibri/plugins/coach/assets/src/modules/coachNotifications/index.js]: I was afraid because I saw that to get the maxId (for the `after` prop) we were using the maxBy, and made me doubt :sweat_smile:. But yes, since we are sorting them in the `INSERT_NOTIFICATIONS` action, they are already sorted. I will change this

File: kolibri/plugins/coach/assets/src/views/common/notifications/ActivityList.vue (hunk 5 above)
Comment (AlexVelezLl) [reply kolibri/plugins/coach/assets/src/views/common/notifications/ActivityList.vue:158]: The `before` param is already set inside the `moreNotificationsForClass` action https://github.com/AlexVelezLl/kolibri/blob/cdb559e453c0c6468ef95d4c97ad7bfdb342d9ec/kolibri/plugins/coach/assets/src/modules/coachNotifications/index.js#L104. So it was repeated code and thought it was better to keep the one in the `moreNotificationsForClass` action,

File: kolibri/plugins/coach/api.py (hunk 2 above)
Comment (rtibbles) [reply kolibri/plugins/coach/api.py]: You should just be able to pass a Javascript Date object as a parameter to the request, and it should get properly serialized.

We may want to update the filtering that is happening here to use a django_filter Filterset (similarly to what we do in other places) as this will handle properly turning the appropriate filter values into the right sort of objects.

For filtering on a datetime, we would use the DateTimeFilter: https://django-filter.readthedocs.io/en/stable/ref/filters.html#datetimefilter

Unfortunately, there aren't any examples of using this specific filter in the codebase, but for example you would use a field definition like this:

```
before = DateTimeFilter(field_name="timestamp", lookup_expr="lt")
after = DateTimeFilter(field_name="timestamp", lookup_expr="gt")
```

For a fuller example of using a FilterSet, see here: https://github.com/learningequality/kolibri/blob/2692958f7772b099cf9191ff47e685ff4ca83206/kolibri/core/content/api.py#L421

(you shouldn't have to use methods for most of your filtering)

File: kolibri/core/notifications/test/test_api.py [hunk 6]
```diff
@@ -74,7 +75,6 @@ def setUp(self):
         self.contentnode_ids = [
             "25f32edcec565396a1840c5413c92451",
             "25f32edcec565396a1840c5413c92452",
-            "25f32edcec565396a1840c5413c92453",
```
Comment (rtibbles) [inline kolibri/core/notifications/test/test_api.py:77]: I assume this was just for simplicity of not having to make so many summary logs for completion?

File: kolibri/core/notifications/test/test_api.py [hunk 7]
```diff
@@ -168,22 +200,232 @@ def test_create_notification(self):
...
+    ):
+        self.summarylog1.progress = 1.0
+        self.summarylog1.completion_timestamp = local_now() + timedelta(seconds=2)
+        self.summarylog1.end_timestamp = local_now() + timedelta(seconds=5)
+        self.summarylog1.save()
+        parse_summarylog(self.summarylog1)
+
+        # Student retried the exercise, and even with progress = 0, this should
+        # be considered as completed
+        self.summarylog1.progress = 0
```
Comment (rtibbles) [inline kolibri/core/notifications/test/test_api.py:297]: Hrm - I feel like this should only be the case when we can also independently confirm that there's a masterylog that does have the completion registered?

Would be good to duplicate this test as is, and assert that it doesn't set the lesson as complete - and then in the other copy add a complete and an incomplete masterylog, and assert that it does set the lesson as complete.

File: kolibri/core/notifications/test/test_api.py [hunk 8]
```diff
@@ -168,22 +200,232 @@ def test_create_notification(self):
...
+            complete=True,
+        )
+
+        parse_summarylog(self.summarylog1)
+
+        self.summarylog2.progress = 1.0
+        self.summarylog2.completion_timestamp = local_now() + timedelta(seconds=2)
+        self.summarylog2.end_timestamp = local_now() + timedelta(seconds=5)
+        self.summarylog2.save()
+        masteryLog_2 = MasteryLogFactory.create(
```
Comment (rtibbles) [inline kolibri/core/notifications/test/test_api.py]: This variable name is mildly confusing - as calling it `masteryLog_2` makes it feel like it's the second masterylog for something (although I see it is meant to be the masterylog for summarylog2).

Maybe just keeping the casing consistent with the above would help, so `masterylog2` to parallel `sumamrylog2` (that would also remove the little diversion into camelCase too :) )

File: kolibri/core/notifications/test/test_api.py [hunk 9]
```diff
@@ -756,371 +876,261 @@ def test_parse_attemptslog_update_attempt_with_three_wrong_attempts(
             lesson_id=self.lesson_id,
             contentnode_id=self.node_1.id,
             reason=HelpReason.Multiple,
-            timestamp=attemptlog3.start_timestamp,
+            timestamp=attemptlog3.end_timestamp,
         )
 
     @patch("kolibri.core.notifications.api.create_notification")
     @patch("kolibri.core.notifications.api.save_notifications")
-    def test_update_lesson_assessment_with_three_wrong_attempts_on_same_attempt(
```
Comment (rtibbles) [inline kolibri/core/notifications/test/test_api.py:764]: Why the change from three to four? Three is the current expected behaviour.

File: kolibri/core/notifications/test/test_api.py [hunk 10]
```diff
@@ -756,371 +876,261 @@ def test_parse_attemptslog_update_attempt_with_three_wrong_attempts(
...
-            end_timestamp=now,
-            time_spent=1.0,
-            complete=True,
-            correct=1,
-            hinted=False,
-            error=False,
-            interaction_history=[interactions[0]],
-        )
+        attemptlog = self._create_mock_attemptlog()
+        interactions = [{"type": "answer", "correct": 0} for _ in range(4)]
```
Comment (rtibbles) [inline kolibri/core/notifications/test/test_api.py]: Likewise here.

File: kolibri/core/notifications/test/test_api.py [hunk 11]
```diff
@@ -756,371 +876,261 @@ def test_parse_attemptslog_update_attempt_with_three_wrong_attempts(
...
-            end_timestamp=now,
-            time_spent=1.0,
-            complete=True,
-            correct=0,
-            hinted=False,
-            error=False,
-            interaction_history=interactions,
-        )
+        attemptlog = self._create_mock_attemptlog()
+        interactions = [{"type": "answer", "correct": 0} for _ in range(4)]
```
Comment (rtibbles) [inline kolibri/core/notifications/test/test_api.py]: And here.

File: kolibri/core/notifications/test/test_api.py [hunk 12]
```diff
@@ -1131,137 +1141,141 @@ def setUpTestData(cls):
         cls.facility = FacilityFactory.create()
         cls.superuser = create_superuser(cls.facility)
 
-        cls.user1 = FacilityUserFactory.create(facility=cls.facility)
-        cls.user2 = FacilityUserFactory.create(facility=cls.facility)
+    def setUp(self):
```
Comment (rtibbles) [inline kolibri/core/notifications/test/test_api.py]: Hrm, what was the need for this update? Setting up commonly used initial data in the `setUpTestData` classmethod helps tests to run considerably faster, so if we can keep this at all, we should.

**Review by rtibbles — DISMISSED** [review]
A few more things that seem important here:
* Clearly defining the role of the masterylogs for the retry behaviour completion notification generation
* Reverting the change from 3 incorrect attempts on the same question to 4
* Reverting the change from setUpTestData to setUp if at all possible

One tiny nit about variable name/casing - but this is not blocking.

File: kolibri/core/notifications/test/test_api.py (hunk 6 above)
Comment (AlexVelezLl) [reply kolibri/core/notifications/test/test_api.py:77]: Hmm it was because this value wasnt being used anywhere, we are creating/reading just 2 nodes (I thought I was reverted the deletion tho :sweat_smile:)

File: kolibri/core/notifications/test/test_api.py (hunk 9 above)
Comment (AlexVelezLl) [reply kolibri/core/notifications/test/test_api.py:764]: Oh, its because the help needed notification jumps when there are _more_ than 3 failed interaction (it was like that even before the refactor https://github.com/learningequality/kolibri/blob/c445798fad324c81fa076c78e1fb95b20ad6cb91/kolibri/core/notifications/api.py#L695). For now having `[{"type": "answer", "correct": 0} for _ in range(3)]` wasnt a problem, because we were creating more attemptlogs in the setup method, and we ended with four or more failed interactions in total.

So this test name was misleading because of that, but now I wonder what the spec is, and if the help needed really should jump with 3 or more failed interactions and what is wrong is the production code :thinking: (Even though the code have been like that for more than 6 years).

Comment (rtibbles) [reply kolibri/core/notifications/test/test_api.py:764]: Hrm, an excellent question. I was definitely always under the impression that it was for 3 or more, not more than three, I will do some digging and see if there is anything to back that up!

File: kolibri/core/notifications/test/test_api.py (hunk 6 above)
Comment (rtibbles) [reply kolibri/core/notifications/test/test_api.py:77]: Ah, if it was never used, no worries about clearing it up, just wanted to check why!

File: kolibri/core/notifications/test/test_api.py (hunk 7 above)
Comment (AlexVelezLl) [reply kolibri/core/notifications/test/test_api.py:297]: Oh, here I'm relying more on the assumption that if the completion_timestamp has a value (we set it [here](https://github.com/AlexVelezLl/kolibri/blob/39c612040566b612fd0bef4ca3d47d90e17d5d2e/kolibri/core/notifications/test/test_api.py#L284)), it means that at some point it was completed, and if it has progress 0 and completion_timestamp != null, then it's because there was a retry.

File: kolibri/core/notifications/test/test_api.py (hunk 12 above)
Comment (AlexVelezLl) [reply kolibri/core/notifications/test/test_api.py]: I'm pretty sure at some point it didn't work with `setUpTestData` and that's why I changed it to the `setUp` method, but I just reverted it and it works perfectly lol haha. Done!

File: kolibri/core/notifications/test/test_api.py (hunk 8 above)
Comment (AlexVelezLl) [reply kolibri/core/notifications/test/test_api.py]: Done!

File: kolibri/core/notifications/test/test_api.py (hunk 12 above)
Comment (AlexVelezLl) [reply kolibri/core/notifications/test/test_api.py]: I also updated the `NotificationsAPITestCase` class to use the `setUpTestData` instead of `setUp`.

File: kolibri/core/notifications/test/test_api.py (hunk 7 above)
Comment (AlexVelezLl) [reply kolibri/core/notifications/test/test_api.py:297]: I updated the test to include the masterylog to make it more attached to reality, but I did not add the assert test that a lesson notification was not generated with an uncompleted resource, because that would be being tested [here](https://github.com/AlexVelezLl/kolibri/blob/39c612040566b612fd0bef4ca3d47d90e17d5d2e/kolibri/core/notifications/test/test_api.py#L216).

**Comment by AlexVelezLl** [comment]
Thank you @rtibbles! I have updated the PR :)

File: kolibri/core/notifications/test/test_api.py (hunk 7 above)
Comment (rtibbles) [reply kolibri/core/notifications/test/test_api.py:297]: OK, we can keep this as is for now, I think I need to mull further on what the implications for this might be - but it's not a regression by any means!

**Review by rtibbles — COMMENTED** [review]
This looks good to go from the code side @radinamatic if you could do final QA to confirm that the issue you reported is fixed, then we can merge this.

**Comment by pcenov** [comment]
Hi @AlexVelezLl @radinamatic - the only issue I was able to identify while testing this is the following error in the console, with seemingly no user facing consequences but I'm going to mention it to be on the safe side:

`TypeError: Cannot read properties of undefined (reading 'correct')`



Logs: [LogsAndDB.zip](https://github.com/user-attachments/files/16542506/LogsAndDB.zip)

Other than that everything else is working properly and I am seeing all of the notifications.

**Comment by rtibbles** [comment]
Looking at the component in question, this looks like an attemptLog was somehow undefined? https://github.com/learningequality/kolibri/blob/develop/kolibri/core/assets/src/views/AttemptLogItem.vue#L83

This is _probably_ not caused by the changes here - @AlexVelezLl if you feel like chasing this down, please do, otherwise we can file a follow up issue.

**Comment by AlexVelezLl** [comment]
Im gonna give a try @rtibbles. If it becomes something too big we can file a follow up issue :)

**Comment by AlexVelezLl** [comment]
I did a little research, and this issue is happening in the ExamReport page of exercise resources because in the AttemptLogList we [iterate through the section.questions](https://github.com/AlexVelezLl/kolibri/blob/17ab6f78c8808dbd7ae2f40b49b9b16a1582fc03/kolibri/core/assets/src/views/AttemptLogList.vue#L112), but for a practice resource we dont have attemptLogs for every question, since we dont necessarily have attempted every question, so that why this attemptLog is undefined here https://github.com/AlexVelezLl/kolibri/blob/17ab6f78c8808dbd7ae2f40b49b9b16a1582fc03/kolibri/core/assets/src/views/AttemptLogList.vue#L135.

Im not sure what the order should be in the exercises attemptLog  and how should it be iterated. So it will be better to file a new issue to solve Exam Reports of exercise resources @rtibbles @pcenov @radinamatic.
image

**Comment by rtibbles** [comment]
Thanks, @AlexVelezLl - seems like this is likely a regression from the EQM work.

**Review by rtibbles — APPROVED** [review]
Unrelated regression has been filed as a follow up. This is good to go!
